node/metered-channel/src/oneshot.rs

```diff
@@ -0,0 +1,411 @@
+	/// Access the measurements
+	pub fn measurements(&self) -> Measurements {
+		self.measurements.clone()
+	}
+
+	/// Converts the wrapper type into it's inner value.
+	///
+	/// `trait Into` cannot be implemented due to conflicts.
+	pub fn into(self) -> T {
+		self.value
+	}
+}
+
+impl<T> AsRef<T> for OutputWithMeasurements<T> {
+	fn as_ref(&self) -> &T {
+		&self.value
+	}
+}
+
+impl<T> Deref for OutputWithMeasurements<T> {
+	type Target = T;
+
+	fn deref(&self) -> &Self::Target {
+		&self.value
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use assert_matches::assert_matches;
+	use futures::{executor::ThreadPool, task::SpawnExt};
+
+	use super::*;
+
+	#[derive(Clone, PartialEq, Eq, Debug)]
+	struct DummyItem {
+		vals: [u8; 256],
+	}
+
+	impl Default for DummyItem {
+		fn default() -> Self {
+			Self { vals: [0u8; 256] }
+		}
+	}
+
+	fn test_launch<S, R, FS, FR>(name: &'static str, gen_sender_test: S, gen_receiver_test: R)
+	where
+		S: Fn(MeteredSender<DummyItem>) -> FS,
+		R: Fn(MeteredReceiver<DummyItem>) -> FR,
+		FS: Future<Output = ()> + Send + 'static,
+		FR: Future<Output = ()> + Send + 'static,
+	{
+		let _ = env_logger::builder().is_test(true).filter_level(LevelFilter::Trace).try_init();
+
+		let pool = ThreadPool::new().unwrap();
+		let (tx, rx) = channel(name, Duration::from_secs(1), Duration::from_secs(3));
+		futures::executor::block_on(async move {
+			let handle_receiver = pool.spawn_with_handle(gen_receiver_test(rx)).unwrap();
+			let handle_sender = pool.spawn_with_handle(gen_sender_test(tx)).unwrap();
+			futures::future::select(
+				futures::future::join(handle_sender, handle_receiver),
+				Delay::new(Duration::from_secs(5)),
+			)
+			.await;
+		});
+	}
+
+	use log::LevelFilter;
+
+	#[test]
+	fn easy() {
+		test_launch(
+			"easy",
+			|tx| async move {
+				tx.send(DummyItem::default()).unwrap();
+			},
+			|rx| async move {
+				let x = rx.await.unwrap();
+				let measurements = x.measurements();
+				assert_eq!(x.as_ref(), &DummyItem::default());
+				dbg!(measurements);
+			},
+		);
+	}
+
+	#[test]
+	fn cancel_by_drop() {
+		test_launch(
+			"cancel_by_drop",
+			|tx| async move {
+				Delay::new(Duration::from_secs(2)).await;
+				drop(tx);
+			},
+			|rx| async move {
+				let result = rx.await;
+				assert_matches!(result, Err(Error::Canceled(_, _)));
+				dbg!(result.measurements());
+			},
+		);
+	}
+
+	#[test]
+	fn starve_till_hard_timeout() {
+		test_launch(
+			"starve_till_timeout",
+			|tx| async move {
+				Delay::new(Duration::from_secs(4)).await;
+				let _ = tx.send(DummyItem::default());
+			},
+			|rx| async move {
+				let result = rx.await;
+				assert_matches!(&result, e @ &Err(Error::HardTimeout(_, _)) => {
+					println!("{:?}", e);
+				});
+				dbg!(result.measurements());
+			},
+		);
+	}
+
+	#[test]
+	fn starve_till_soft_timeout_then_food() {
+		test_launch(
+			"starve_till_soft_timeout_then_food",
+			|tx| async move {
+				Delay::new(Duration::from_secs(2)).await;
```

Comment:
Even though it's just a test, I think it would be better to make this a constant.

Reply:
They are all used exactly once, and are really only relevant per test case (and each of them is super short), so I don't think that's necessary (though I generally agree with you).